File: src/deepview_canvas/widgets/canvas_list/src/widget.ts
/**
 * Copyright 2024 BetterWithData
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License. 
 */

import { MODULE_NAME, MODULE_VERSION } from './version';
import {
  DOMWidgetModel,
  ISerializers,
} from '@jupyter-widgets/base';

import {serializers, CanvasBaseDomWidgetView } from '@betterwithdata/canvas_viz';

import CanvasList from './CanvasList.svelte';

export class CanvasListModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: CanvasListModel.model_name,
      _model_module: CanvasListModel.model_module,
      _model_module_version: CanvasListModel.model_module_version,
      _view_name: CanvasListModel.view_name,
      _view_module: CanvasListModel.view_module,
      _view_module_version: CanvasListModel.view_module_version,
    };
  }
  static serializers: ISerializers = serializers;

  static model_name = 'CanvasListModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'CanvasListView'; // Set to null if no view
  static view_module = MODULE_NAME; // Set to null if no view
  static view_module_version = MODULE_VERSION;
}

export class CanvasListView extends CanvasBaseDomWidgetView {
  render() {
    this.setupWidget(this.model, CanvasList);
  }
}

File: src/deepview_canvas/widgets/canvas_list/canvas_list/widget.py
#
# Copyright 2024 BetterWithData
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from pathlib import Path
import canvas_ux
from traitlets import Unicode
from typing import Tuple
from ._frontend import module_name, module_version


class CanvasList(canvas_ux.CanvasWidget):
    _model_name = Unicode(
        'CanvasListModel').tag(sync=True)  # type: ignore
    _view_name = Unicode(
        'CanvasListView').tag(sync=True)  # type: ignore
    _model_module = Unicode(module_name).tag(sync=True)  # type: ignore
    _model_module_version = Unicode(module_version).tag(sync=True)  # type: ignore
    _view_module = Unicode(module_name).tag(sync=True)  # type: ignore
    _view_module_version = Unicode(module_version).tag(sync=True)  # type: ignore
    name = 'CanvasList'
    description = "A Canvas component that displays a view of data instances"

    def __init__(self,
                 width: str = 'XXL',
                 height: str = 'M',
                 page: str = 'List',
                 **kwargs: dict
                 ):
        """A Canvas component that displays a view of data instances

        Parameters
        ----------
        width : str, optional
            By default "XXL".
        height : str, optional
            By default "M".
        page : str, optional
            Which page of the dashboard to show the widget on, by default "CanvasList"

        Returns
        -------
        CanvasList
        """
        super().__init__(**kwargs)
        self.width: str = width
        self.height: str = height
        self.page: str = page
        self.spec = canvas_ux.WidgetSpec(
            width=width,
            height=height,
            page=page,
            name=self.name,
            description=self.description
        )

    def js_path(self) -> Tuple[Path, Path]:
        app_files_path = Path(
            (Path(__file__).parent), 'standalone', 'widgets')
        js_path = Path(app_files_path, 'CanvasList.js')
        map_path = Path(
            app_files_path, 'CanvasList.js.map')
        return js_path, map_path
